### PR TITLE
Implement Huggingface model download in storage initializer

### LINKED ISSUE #3545: Support download huggingface models in storage initializer
/kind feature

**Describe the solution you'd like**

I would like to use storage initializer to pull the huggingface [models](https://huggingface.co/models) which fits the KServe experience. StorageUri will be in this format: `hf://<repo>/<model>:<hash(optional)>`

```yaml
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: bloom
spec:
  predictor:
    model:
      modelFormat:
        name: huggingface
      #runtime: kserve-huggingface
      storageUri: hf://bigscience/bloom-560m:<hash>
      resources:
        limits:
          cpu: "2"
          memory: 4Gi
        requests:
          cpu: "1"
          memory: 2Gi
```

**Anything else you would like to add:**

Another option is to support any `Git LFS`.

**Links to the design documents:**
[Optional, start with the short-form [RFC template](https://docs.google.com/document/d/1UcBeLfZ_JMGpVrPJmYtEIVH_9Y4U3AEKQdq_IKuOMrU) to outline your ideas and get early feedback.]
[Required, use the longer-form [design doc template](https://docs.google.com/document/d/1Mtoui_PP2a9N59NjYHnsvrdJ8t2iKFwIJAx1zRO_I1c) to specify and discuss your design in more detail]

### PR DESCRIPTION
**Pull request by andyi2it**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3545 

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: python/kserve/kserve/storage/storage.py [hunk 1]
```diff
@@ -272,8 +271,28 @@ def _download_s3(uri, temp_dir: str):
             if mimetype in ["application/x-tar", "application/zip"]:
                 Storage._unpack_archive_file(target, mimetype, temp_dir)
 
+    @staticmethod
+    def _download_hf(uri, temp_dir: str):
+        from transformers import AutoTokenizer, AutoConfig, AutoModel
```
Comment (lizzzcai) [inline python/kserve/kserve/storage/storage.py]: instead of loading model and then saving the model to local, which means you need resources (lot of memory) to load the model, maybe [snapshot_download](https://github.com/vllm-project/vllm/blob/65bf2ac165734fb6339210c4b2b8ce68d2391b77/vllm/model_executor/model_loader/weight_utils.py#L203-L210) SDK will be better.

File: .github/workflows/e2e-test.yml [hunk 2]
```diff
@@ -24,6 +24,7 @@ env:
   # Controller images
   CONTROLLER_IMG: "kserve-controller"
   STORAGE_INIT_IMG: "storage-initializer"
+  HF_STORAGE_INIT_IMG: "hf-storage-initializer"
```
Comment (lizzzcai) [inline .github/workflows/e2e-test.yml]: why have a new image here? I assume the feature will be in the storage initializer and we should use the single image for all types of storage.

Comment (andyi2it) [reply .github/workflows/e2e-test.yml]: @lizzzcai 
The problem is storage initializer image size grows significantly to support huggingface server, from 500+ Mb to 1.2+ Gb approximately. That would be fairly large increase in image size for a feature only a few subset of users may use.
So, we decided to use a separate image for huggingface server.

File: config/storagecontainers/default.yaml [hunk 3]
```diff
@@ -11,7 +11,7 @@ spec:
         memory: 100Mi
         cpu: 100m
       limits:
-        memory: 1Gi
+        memory: 3Gi
```
Comment (yuzisun) [inline config/storagecontainers/default.yaml]: any reason to bump up to 3Gi?

File: docs/samples/storage/hf/hf_isvc.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  predictor:
+    serviceAccountName: hfserviceacc
+    model:
+      modelFormat:
+        name: huggingface
+      args:
+      - --model_name=llama2
+      - --model_dir=hf://meta-llama/Llama-2-7b-chat-hf
+      - --tensor_input_names=input_ids
+      - --disable_vllm
```
Comment (yuzisun) [inline docs/samples/storage/hf/hf_isvc.yaml]: this flag is already removed and replaced by --backend

File: docs/samples/storage/hf/hf_isvc.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  name: huggingface-llama2
+spec:
+  predictor:
+    serviceAccountName: hfserviceacc
+    model:
+      modelFormat:
+        name: huggingface
+      args:
+      - --model_name=llama2
+      - --model_dir=hf://meta-llama/Llama-2-7b-chat-hf
```
Comment (yuzisun) [inline docs/samples/storage/hf/hf_isvc.yaml]: This should be using `storageUri` field

File: python/kserve/kserve/storage/storage.py [hunk 6]
```diff
@@ -154,6 +148,9 @@ def _update_with_storage_spec():
 
     @staticmethod
     def get_S3_config():
+        from botocore import UNSIGNED
+        from botocore.client import Config
+
```
Comment (yuzisun) [inline python/kserve/kserve/storage/storage.py:159]: I think maybe better to separate them out into different files now?

File: python/hf-storage-initializer.Dockerfile [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+# Activate virtual env
+ARG VENV_PATH
+ENV VIRTUAL_ENV=${VENV_PATH}
+RUN python3 -m venv $VIRTUAL_ENV
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
+
+COPY kserve/pyproject.toml kserve/poetry.lock kserve/
+RUN cd kserve && poetry install --no-root --no-interaction --no-cache --extras "hf-storage"
+COPY kserve kserve
```
Comment (yuzisun) [inline python/hf-storage-initializer.Dockerfile]: does storage initializer actually need the kserve dependency?

File: python/kserve/kserve/storage/test/test_storage.py [hunk 8]
```diff
@@ -235,17 +235,16 @@ def test(_):
     mock.patch("requests.get", return_value=response)(test)()
 
 
-@mock.patch(STORAGE_MODULE + ".storage")
-def test_mock_gcs(mock_storage):
+@mock.patch("google.cloud.storage.Client")
```
Comment (yuzisun) [inline python/kserve/kserve/storage/test/test_storage.py]: this test should be in `test_gcs_storage.py`

File: python/kserve/kserve/storage/test/test_storage.py [hunk 9]
```diff
@@ -235,17 +235,16 @@ def test(_):
...
+def test_mock_gcs(mock_client):
```
Comment (yuzisun) [inline python/kserve/kserve/storage/test/test_storage.py]: same here

File: python/kserve/kserve/storage/test/test_storage.py [hunk 10]
```diff
@@ -255,21 +254,21 @@ def test_storage_blob_exception():
         Storage.download(blob_path)
 
 
-@mock.patch(STORAGE_MODULE + ".boto3")
-def test_storage_s3_exception(mock_boto3):
+@mock.patch("boto3.resource")
+def test_storage_s3_exception(mock_resource):
```
Comment (yuzisun) [inline python/kserve/kserve/storage/test/test_storage.py]: this test should be in `test_s3_storage.py`

File: python/kserve/kserve/storage/test/test_storage.py [hunk 11]
```diff
@@ -303,3 +302,29 @@ def test_unpack_zip_file():
     Storage._unpack_archive_file(tar_file, mimetype, out_dir)
     assert os.path.exists(os.path.join(out_dir, "model.pth"))
     os.remove(os.path.join(out_dir, "model.pth"))
+
+
+def test_download_hf():
```
Comment (yuzisun) [inline python/kserve/kserve/storage/test/test_storage.py]: move this test to `test_hf_storage.py`

File: python/kserve/pyproject.toml [hunk 12]
```diff
@@ -61,6 +66,8 @@ azure-storage-blob = { version = "^12.10.0", optional = true }
 azure-storage-file-share = { version = "^12.7.0", optional = true }
 azure-identity = { version = "^1.8.0", optional = true }
 boto3 = { version = "^1.21.0", optional = true }
+transformers = { version = "^4.37.0", optional = true }
+torch = { version = "^2.3.0+cpu", optional = true, source = "pytorch"}
```
Comment (yuzisun) [inline python/kserve/pyproject.toml]: I think the hf storage can be standalone and does not need to depend on kserve package

File: .github/workflows/e2e-test.yml (hunk 2 above)
Comment (lizzzcai) [reply .github/workflows/e2e-test.yml]: Hi @andyi2it , to just download huggingface repo, I think [huggingface_hub](https://github.com/huggingface/huggingface_hub) should be enough, and the size should be acceptable. In this case, I think we don't need to split the image.

File: python/kserve/kserve/storage/storage.py (hunk 1 above)
Comment (lizzzcai) [reply python/kserve/kserve/storage/storage.py]: Hi @andyi2it , as mentioned above, [snapshot_download](https://huggingface.co/docs/huggingface_hub/v0.24.2/en/package_reference/file_download#huggingface_hub.snapshot_download) sdk will be the better option to download the model artifacts.

File: python/kserve/kserve/storage/storage.py [hunk 13]
```diff
@@ -278,8 +277,29 @@ def _download_s3(uri, temp_dir: str) -> str:
...
+        components = uri.split("://")[1].split("/")
+
+        repo = components[0]
+        model, _, hash_value = components[1].partition(":")
+        revision = hash_value if hash_value else None
+
+        tokenizer = AutoTokenizer.from_pretrained(f"{repo}/{model}", revision=revision)
+        model_config = AutoConfig.from_pretrained(f"{repo}/{model}", revision=revision)
+        model = AutoModel.from_config(model_config)
```
Comment (juhyeon-cha) [inline python/kserve/kserve/storage/storage.py]: custom code models must set the trust_remote_code argument to true.
need to use either the inference service's model arguments or the HF hub model API.

File: docs/samples/storage/hf/hf_secret.yaml [hunk 14]
```diff
@@ -0,0 +1,14 @@
+apiVersion: v1
```
Comment (spolti) [inline docs/samples/storage/hf/hf_secret.yaml:1]: isn't better to call this file `hf_storage_config.yaml`?

File: pkg/credentials/hf/hf_secret.go [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+const (
+	HFTokenKey = "HF_TOKEN"
+)
+
+func BuildSecretEnvs(secret *v1.Secret) []v1.EnvVar {
+	envs := make([]v1.EnvVar, 0)
+
+	if token, ok := secret.Data[HFTokenKey]; ok {
```
Comment (spolti) [inline pkg/credentials/hf/hf_secret.go:30]: Isn't this env mandatory?
Maybe we could print a warning or even panic.

File: python/hf-storage-initializer.Dockerfile [hunk 16]
```diff
@@ -0,0 +1,52 @@
+ARG PYTHON_VERSION=3.9
+ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
+ARG VENV_PATH=/prod_venv
+
+FROM ${BASE_IMAGE} as builder
+
+# Install Poetry
+ARG POETRY_HOME=/opt/poetry
+ARG POETRY_VERSION=1.7.1
```
Comment (spolti) [inline python/hf-storage-initializer.Dockerfile]: I think we should use 1.8.2

File: python/storage-initializer.Dockerfile [hunk 17]
```diff
@@ -48,6 +48,8 @@ ENV PATH="$VIRTUAL_ENV/bin:$PATH"
 
 RUN useradd kserve -m -u 1000 -d /home/kserve
 
+RUN apt-get -y update && apt-get install ncdu
```
Comment (spolti) [inline python/storage-initializer.Dockerfile]: why is this needed?

File: python/hf-storage-initializer.Dockerfile (hunk 7 above)
Comment (spolti) [reply python/hf-storage-initializer.Dockerfile]: a question, can't the existing storage initializer image be reused?
Thinking that users would need to switch manually the initializer images, if so, how should it be done?

Comment (andyi2it) [reply python/hf-storage-initializer.Dockerfile]: By default the non hf storage initializer's image will be pulled. If the storage uri contains hf prefix the hf-storage-initializer image will be pulled. The end user needn't switch manually.

The reason why we have 2 images is because hf storage initializer image is twice the size.

Comment (andyi2it) [reply python/hf-storage-initializer.Dockerfile]: @spolti ☝️

File: docs/samples/storage/hf/hf_secret.yaml (hunk 14 above)
Comment (andyi2it) [reply docs/samples/storage/hf/hf_secret.yaml:1]: We were following the same sample pattern as s3.

File: pkg/credentials/hf/hf_secret.go (hunk 15 above)
Comment (andyi2it) [reply pkg/credentials/hf/hf_secret.go:30]: I suppose this secret is mandatory only in case of pulling private HuggingFace repository.

File: python/hf-storage-initializer.Dockerfile (hunk 7 above)
Comment (lizzzcai) [reply python/hf-storage-initializer.Dockerfile]: is it possible to reduce the image size by using some other light-weighted sdk? adding a new image means the operation team needs to maintain a new image(security scan etc) just for models from HF.

Comment (spolti) [reply python/hf-storage-initializer.Dockerfile]: okay, thanks.

Comment (spolti) [reply python/hf-storage-initializer.Dockerfile]: btw, I am working on this https://github.com/kserve/kserve/issues/3489
it is related to it.

the plan is to decouple the storage package from KServe and make a hard copy of the logging.py file, so the storage-initializer can be completely independent from KServe avoiding importing/installing a few unneeded dependencies, like ray and others.
Like

File: python/kserve/kserve/storage/storage.py [hunk 18]
```diff
@@ -284,8 +283,26 @@ def _download_s3(uri, temp_dir: str) -> str:
...
+        from huggingface_hub import snapshot_download
+
+        components = uri.split("://")[1].split("/")
+
+        repo = components[0]
+        model, _, hash_value = components[1].partition(":")
+        revision = hash_value if hash_value else None
+
+        snapshot_download(
+            repo_id=f"{repo}/{model}", revision=revision, local_dir=temp_dir
```
Comment (lizzzcai) [inline python/kserve/kserve/storage/storage.py:314]: HF repo usually contains multiple formats, for example, `safetensors`, `bin` etc. Most of the case, only one should be downloaded to reduce the download time (should save half of the time). 

Usually I will add the allow/ignore patterns, for example:

```python
# here safetensors is the preferred format.
allow_patterns = ["*.json", "*.safetensors", "*.model"]
ignore_patterns = ["*.msgpack", "*.h5", "*.bin"]

output_dir = snapshot_download(
      repo_id=model_name,
      revision=revision,
      allow_patterns=allow_patterns,
      ignore_patterns=ignore_patterns,
      local_dir=models_path,
)
```
I think it is quite needed, but I am not sure the best way to pass it, for example via the `ALLOW_PATTERNS` and `IGNORE_PATTERNS` in the secret?

**Comment by lizzzcai** [comment]
/lgtm
/assign @yuzisun 
the initial support for hf model download looks good to me. for the allow/ignore patterns, can be done in another PR.

File: python/kserve/kserve/storage/storage.py [hunk 19]
```diff
@@ -284,8 +283,26 @@ def _download_s3(uri, temp_dir: str) -> str:
                 temp_dir = Storage._unpack_archive_file(target, mimetype, temp_dir)
         return temp_dir
 
+    @staticmethod
+    def _download_hf(uri, temp_dir: str) -> str:
+        from huggingface_hub import snapshot_download
+
+        components = uri.split("://")[1].split("/")
```
Comment (yuzisun) [inline python/kserve/kserve/storage/storage.py]: @andyi2it Can you add a check here to validate the hf url  ?

File: python/kserve/kserve/storage/test/test_hf_storage.py [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest.mock as mock
+
+from kserve.storage import Storage
+
+
+@mock.patch("huggingface_hub.snapshot_download")
+def test_download_model(mock_snapshot_download):
```
Comment (yuzisun) [inline python/kserve/kserve/storage/test/test_hf_storage.py:22]: Add a test for invalidate hf url

**Review by yuzisun — APPROVED** [review]
